### PR TITLE
Don't GC unit if it is in cgroup_empty_queue

### PR DESCRIPTION
**Pull request by rphibel**
The gc_unit_queue is dispatched before the cgroup_empty_queue.  Because
of this, when we enter in `on_cgroup_empty_event`, the unit in
`cgroup_empty_queue` may already have been freed and we don't clean up the
corresponding cgroup. With this change, we prevent the unit from being
garbage collected if it is in the `cgroup_empty_queue`.

### CODE REVIEW

**Comment by rphibel** [comment]
@poettering Could you please have a look? It seems that it solves the issue of slice cgroups not being cleaned up. But maybe there are corner cases when this approach could cause an issue.

File: test/units/testsuite-82.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
```
Comment (yuwata) [inline test/units/testsuite-82.sh]: Please do not create new TEST-82, but add the test as a sub test of TEST-19. That is, please rename testsuite-82.sh as testsuite-19.cleanup-slice.sh or so.

Note, TEST-82 is also will be used #27435.

File: src/core/manager.c [hunk 2]
```diff
@@ -1250,6 +1250,31 @@ static unsigned manager_dispatch_cleanup_queue(Manager *m) {
...
+                (void) unit_check_oomd_kill(u);
+
+                unit_add_to_gc_queue(u);
+
+                if (UNIT_VTABLE(u)->notify_cgroup_empty)
+                        UNIT_VTABLE(u)->notify_cgroup_empty(u);
+        }
+
+        return n;
+}
```
Comment (poettering) [inline src/core/manager.c]: this looks wrong. now there are two places where the cgroup empty queue is dispatched, here, synchronously from the event loop, and very similar in style from a low-prio defer event source. the low-prefere defer event source exists so that we can let it run *after* the SIGCHLD handler.

If you want to ensure the GC logic runs only after cgroup empty events have been handled for a unit, then modify unit_may_gc() maybe to make sure that units are never GC'ed which are in the cgroup empty event queue.

File: src/core/slice.c [hunk 3]
```diff
@@ -413,6 +413,26 @@ static bool slice_can_freeze(Unit *s) {
...
+        /* If the cgroup empty notification comes when the unit is not active, we must have failed to clean
+         * up the cgroup earlier and should do it now. */
+        case SLICE_DEAD:
+                unit_prune_cgroup(u);
+                break;
+
+        default:
+                ;
+        }
+}
```
Comment (poettering) [inline src/core/slice.c]: hmm, i scope and service units have this already, and i figure all units with cgroups should really have this.

Hence I think we should really make this generic. i.e. move this to on_cgroup_empty_event():

```c
if (UNIT_IS_INACTIVE_OR_DEAD(unit_status(u)) 
        unit_prune_cgroup(u);
```

and then don't even pass control to the per-unit-type handlers for these states, and remove the relevant handling from service/scope/slice code for all states that map to UNIT_DEAD/UNIT_FAILED.

Comment (poettering) [reply src/core/slice.c]: also, this really should be a separate commit in this PR.

File: src/core/manager.c (hunk 2 above)
Comment (poettering) [reply src/core/manager.c]: or in other words, i want that on_cgroup_empty_event() remains the only place where we dispatch the cgroup empty event queue – however, to not have the GC logic interfere, let's make it ineffective if cgroup empty event is pending in the queue for the unit.

File: test/units/testsuite-19.cleanup-slice.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+pkill sleep
+
+assert_rc 1 systemd-cgls /test82.slice/test82.service
+
+# Check that empty cgroup /test82.slice has been removed
+timeout 30 bash -c 'while systemd-cgls /test82.slice >& /dev/null; do sleep .5; done'
+assert_rc 1 systemd-cgls /test82.slice
+
+# End
+touch /testok
```
Comment (yuwata) [inline test/units/testsuite-19.cleanup-slice.sh]: Please drop this line. The file should be created by -19.sh after all sub tests succeed.

File: src/core/manager.c [hunk 5]
```diff
@@ -1299,6 +1299,9 @@ static void unit_gc_sweep(Unit *u, unsigned gc_marker) {
                    GC_OFFSET_GOOD, GC_OFFSET_BAD, GC_OFFSET_UNSURE, GC_OFFSET_IN_PATH))
                 return;
 
+        if (u->in_cgroup_empty_queue)
+                goto good;
+
```
Comment (poettering) [inline src/core/manager.c]: hmm, why place this here rather than in unit_may_gc()?

File: src/core/slice.c (hunk 3 above)
Comment (poettering) [inline src/core/slice.c]: please don't add this here if the following commit then removes it again

File: src/core/cgroup.c [hunk 6]
```diff
@@ -3142,8 +3142,11 @@ static int on_cgroup_empty_event(sd_event_source *s, void *userdata) {
 
         unit_add_to_gc_queue(u);
 
-        if (UNIT_VTABLE(u)->notify_cgroup_empty)
-                UNIT_VTABLE(u)->notify_cgroup_empty(u);
+        if (IN_SET(unit_active_state(u), UNIT_INACTIVE, UNIT_FAILED))
+                unit_prune_cgroup(u);
+        else
+                if (UNIT_VTABLE(u)->notify_cgroup_empty)
```
Comment (poettering) [inline src/core/cgroup.c]: please merge the `else` line and the `if (UNIT_VT…` line into one `else if (UNIT_VT…` line

File: src/core/unit.c [hunk 7]
```diff
@@ -481,6 +481,9 @@ bool unit_may_gc(Unit *u) {
                         return false;
         }
 
+        if (u->in_cgroup_empty_queue)
+                return false;
```
Comment (poettering) [inline src/core/unit.c]: this is a relatively cheap check (i.e. doesn't go to the vfs, doesn't check any other objects), hence please put it somewhere early in the function, maybe right after the check if the unit is perpetual.

otherwise we might do all the expensive work for nothing, and then decide it was all useless after checking the final cheap check...

**Comment by poettering** [comment]
looks great, and beautifully short. if you change the order of the unit_may_gc() checks we are good to go!

File: test/units/testsuite-19.cleanup-slice.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
Comment (poettering) [inline test/units/testsuite-19.cleanup-slice.sh:1]: hmm, this just drops the shell script in to test-19, but it's not actually started, is it? what am i missing?

Comment (rphibel) [reply test/units/testsuite-19.cleanup-slice.sh:1]: `run_subtests` in `testsuite-19.sh` executes all the `testsuite-19.*.sh` subscripts

https://github.com/systemd/systemd/blob/b58feca65cb2426b553828e5dff2eb0b213fa05b/test/units/testsuite-19.sh#L11

**Comment by poettering** [comment]
thanks!

File: test/units/testsuite-19.cleanup-slice.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+set -eux
+set -o pipefail
+
+# shellcheck source=test/units/util.sh
+. "$(dirname "$0")"/util.sh
+
+export SYSTEMD_LOG_LEVEL=debug
+
+# Create service with KillMode=none inside a slice
+cat <<EOF > /run/systemd/system/test19cleanup.service
```
Comment (keszybz) [inline test/units/testsuite-19.cleanup-slice.sh]: Please remove the space after `>`.

File: test/units/testsuite-19.cleanup-slice.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+# Stop slice
+# The sleep process will not be killed because of KillMode=none
+# Since there is still a process running under it, the /test19cleanup.slice cgroup won't be removed
+systemctl stop test19cleanup.slice
+
+# Kill sleep process manually
+pkill sleep
+
+assert_rc 1 systemd-cgls /test19cleanup.slice/test19cleanup.service
```
Comment (keszybz) [inline test/units/testsuite-19.cleanup-slice.sh]: This is racy. Maybe add a sleep or a loop.

**Review by keszybz — CHANGES_REQUESTED** [review]
The tests fail, but I'm not sure what exactly the cause is. The relevant part of the log:
```
May 26 10:39:51 H systemd[1]: Created slice test19cleanup.slice.
May 26 10:39:51 H systemd[1]: Started test19cleanup.service.
May 26 10:39:51 H testsuite-19.sh[511]: + assert_rc 0 systemd-cgls /test19cleanup.slice
May 26 10:39:51 H testsuite-19.sh[518]: + set +ex
May 26 10:39:51 H systemd-cgls[519]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
May 26 10:39:51 H testsuite-19.sh[519]: Control group /test19cleanup.slice:
May 26 10:39:51 H testsuite-19.sh[519]: └─test19cleanup.service
May 26 10:39:51 H testsuite-19.sh[519]:   └─516 sleep infinity
May 26 10:39:51 H testsuite-19.sh[511]: + systemctl stop test19cleanup.slice
May 26 10:39:51 H systemd[1]: test19cleanup.service: Deactivated successfully.
May 26 10:39:51 H systemd[1]: test19cleanup.service: Unit process 516 (4) remains running after unit stopped.
May 26 10:39:51 H systemd[1]: Stopped test19cleanup.service.
May 26 10:39:51 H systemd[1]: Removed slice test19cleanup.slice.
May 26 10:39:51 H testsuite-19.sh[511]: + pkill sleep
May 26 10:39:51 H testsuite-19.sh[422]: + _FAILED_TESTS+=("$subtest")
...
May 26 10:39:53 H testsuite-19.sh[587]: FAILED TESTS:   1:
May 26 10:39:53 H testsuite-19.sh[587]: ------------------
May 26 10:39:53 H testsuite-19.sh[587]: /usr/lib/systemd/tests/testdata/units/testsuite-19.cleanup-slice.sh
May 26 10:39:53 H systemd[1]: testsuite-19.service: Main process exited, code=exited, status=1/FAILURE
May 26 10:39:53 H systemd[1]: testsuite-19.service: Failed with result 'exit-code'.
May 26 10:39:54 H systemd[1]: Failed to start testsuite-19.service.
```

**Comment by YHNdnzj** [comment]
(BTW, please prefix the commit messages with `core:` or so.)

File: test/units/testsuite-19.cleanup-slice.sh (hunk 10 above)
Comment (rphibel) [reply test/units/testsuite-19.cleanup-slice.sh]: Thanks! I saw the failure but couldn't figure out why the test was failing.

File: test/units/testsuite-19.cleanup-slice.sh [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+assert_rc 0 systemd-cgls /test19cleanup.slice
+
+# Stop slice
+# The sleep process will not be killed because of KillMode=none
+# Since there is still a process running under it, the /test19cleanup.slice cgroup won't be removed
+systemctl stop test19cleanup.slice
+
+# Kill sleep process manually
+while pkill sleep
+do
```
Comment (keszybz) [inline test/units/testsuite-19.cleanup-slice.sh]: `while …; do` please, to match what the usual style.

This doesn't really solve the race. We will wait at least 0.1 s after the process disappears, but the cgroup cleanup could take longer than that. Let's see if the tests pass, but even if they do, this seems very fragile. (Unfortunately the CI often runs in very overloaded systems, so any "sane" timeout is not enough. We really need either race-free things or retries or very long timeouts.)

**Comment by keszybz** [comment]
CentOS CI (Arch Linux):
```
19:08:58 Assertion 'r >= 0' failed at src/test/test-execute.c:1268, function prepare_ns(). Aborting.
```
```
19:11:15 [    8.666079] testsuite-19.sh[450]: + pkill sleep
19:11:15 [    8.680801] testsuite-19.sh[450]: + assert_rc 1 systemd-cgls /test19cleanup.slice/test19cleanup.service
19:11:15 [    8.687272] testsuite-19.sh[461]: + set +ex
19:11:15 [    8.690223] testsuite-19.sh[462]: Control group /test19cleanup.slice/test19cleanup.service:
19:11:15 [    8.695695] testsuite-19.sh[462]: └─455 sleep infinity
19:11:15 [    8.699776] testsuite-19.sh[463]: FAIL: expected: '1' actual: '0'
19:11:15 [    8.703934] testsuite-19.sh[375]: + _FAILED_TESTS+=("$subtest")
```

File: test/units/testsuite-19.cleanup-slice.sh (hunk 11 above)
Comment (keszybz) [reply test/units/testsuite-19.cleanup-slice.sh]: Hmm, but looking at this:
```
19:11:15 [    8.666079] testsuite-19.sh[450]: + pkill sleep
19:11:15 [    8.680801] testsuite-19.sh[450]: + assert_rc 1 systemd-cgls /test19cleanup.slice/test19cleanup.service
19:11:15 [    8.687272] testsuite-19.sh[461]: + set +ex
19:11:15 [    8.690223] testsuite-19.sh[462]: Control group /test19cleanup.slice/test19cleanup.service:
19:11:15 [    8.695695] testsuite-19.sh[462]: └─455 sleep infinity
19:11:15 [    8.699776] testsuite-19.sh[463]: FAIL: expected: '1' actual: '0'
```
It seems that `pkill` returns failure? I maybe it's not a race condition that is the problem, but something else.

File: test/units/testsuite-19.cleanup-slice.sh [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+export SYSTEMD_LOG_LEVEL=debug
+
+# Create service with KillMode=none inside a slice
+cat <<EOF >/run/systemd/system/test19cleanup.service
+[Unit]
+Description=Test 19 cleanup Service
+[Service]
+Slice=test19cleanup.slice
+Type=simple
```
Comment (keszybz) [inline test/units/testsuite-19.cleanup-slice.sh]: Please change this to `Type=exec`, so that we have `sleep` actually running when the service is considered started.

**Comment by yuwata** [comment]
Still test-execute failed in TEST-02.

File: src/test/test-execute.c [hunk 13]
```diff
@@ -209,11 +209,19 @@ static bool is_inaccessible_available(void) {
...
 
         assert_se(manager_load_startable_unit_or_warn(m, unit_name, NULL, &unit) >= 0);
         assert_se(unit_start(unit, NULL) >= 0);
+        slice = UNIT_GET_SLICE(unit);
+        if (slice) {
+                /* We need to start the slice as well otherwise the slice cgroup might be pruned
+                 * in on_cgroup_empty_event. */
+                int r = unit_start(slice, NULL);
+                assert_se(r >= 0 || r == -EALREADY);
+        }
```
Comment (poettering) [inline src/test/test-execute.c]: a slice should be started before any contained unit.

please move this up hence. and please make this recursive, just to be fully accurate, i.e. start all parent slice first, up the tree.

**Review by DaanDeMeyer — APPROVED** [review]

**Review by DaanDeMeyer — APPROVED** [review]
